### PR TITLE
Parent and Levels

### PR DESCRIPTION
**Pull request by cshaley**
Added parent attribute functionality for nodes.  Each node has a parent attribute, which is automatically set when defining a node using e.g. `myNode.left = Node(6)`.  Parent definition was also added to other functions in order to incorporate it more broadly.  

Added some basic tree level functionality.  e.g. `myNode.level()` returns distance from the root node.  So for the root node, `level()` returns 0.  `bt.get_levels(t)` returns a dictionary of form `{0:
 [rootnode], 1: [1st level nodes]}`.  

A couple of other small functions are included as well.
Sample functionality:

In:
```
t = convert(list(range(10)))
bt.show_ids(t)
```
Out:
```
        ____0__    
       /       \   
    __1__       2  
   /     \     / \ 
  3       4   5   6
 / \     /         
7   8   9          
```
                   
In:
```
t.right.right.right = Node(11)
bt.show_ids(t)
```
Out:
```
        ____0__       
       /       \      
    __1__       2     
   /     \     / \    
  3       4   5   6   
 / \     /         \  
7   8   9           10
```                
In :
```
t.right.right.right.parent
```
Out:
```
Node(6)
```
In :
```
print(t.parent)
```
Out:
```
None
```
In:
```
t.is_root()
```
Out:
```
True
```
In:
```
t.is_leaf()
```
Out:
```
False
```
In:
```
t.right.right.level()
```
Out:
```
2
```
In:
```
bt.get_levels(t, show_values=True, show_nulls=True)
```
Out:
```
{0: [0], 1: [1, 2], 2: [3, 4, 5, 6], 3: [7, 8, 9, None, None, None, None, 11]}
```
In:
```
bt.get_levels(t, show_values=True)
```
Out:
```
{0: [0], 1: [1, 2], 2: [3, 4, 5, 6], 3: [7, 8, 9, 11]}
```
In:
```
bt.get_level(t,2)
```
Out:
```
[Node(3), Node(4), Node(5), Node(6)]
```

Covers the first three bullets of issue #9

### CODE REVIEW

**Comment by cshaley** [comment]
aw, bummer.  I didn't see your tests file.  I'll hop to it.

File: binarytree/__init__.py [hunk 1]
```diff
@@ -649,6 +685,89 @@ def convert(bt):
...
+    :type level: int
+    :param show_values: whether to convert nodes to values before returning.
+    :type show_values: boolean
+    :param show_nulls: whether to show where empty nodes are.
+    :type show_nulls: boolean
+    :return: dictionary of form {0:[rootnode], 1:[1st, level, nodes]}
+    :rtype: dictionary
+    """
+    if not isinstance(bt, Node):
+        raise ValueError("Expecting a Node.")
```
Comment (joowani) [inline binarytree/__init__.py]: For the user's convenience, before raising an error I would check if the binarytree is in a list representation form. If it is try to convert to tree form first.

File: binarytree/__init__.py [hunk 2]
```diff
@@ -649,6 +685,89 @@ def convert(bt):
...
+            elif show_nulls:
+                nodes.append(_create_node(_null))
+            if node.right:
+                nodes.append(node.right)
+            elif show_nulls:
+                nodes.append(_create_node(_null))
+        if len(nodes) == 0 or all(node.value is _null for node in nodes):
+            raise ValueError("Requested level not present in tree.")
+        levels[currlevel] = nodes
+    if show_values is True:
```
Comment (joowani) [inline binarytree/__init__.py]: you could simply do something like
```python
return [node.value for node in level] if show_values else level
```

File: binarytree/__init__.py [hunk 3]
```diff
@@ -649,6 +685,89 @@ def convert(bt):
...
+    :type show_nulls: boolean
+    :return: dictionary of form {0:[rootnode], 1:[1st, level, nodes]}
+    :rtype: dictionary
+    """
+    if not isinstance(bt, Node):
+        raise ValueError("Expecting a Node.")
+    if not isinstance(level, int) or level < 0:
+        raise ValueError("Requested level must be a non-negative integer.")
+    currlevel = 0
+    levels = {currlevel: [bt]}
```
Comment (joowani) [inline binarytree/__init__.py]: We do not need to keep track of all levels (only the last level). We also do not have to create nodes for nulls on the way. I believe this loop can be greatly simplified. You can check out how other methods are traversing down trees level by level.

File: binarytree/__init__.py [hunk 4]
```diff
@@ -649,6 +685,89 @@ def convert(bt):
...
+                nodes.append(_create_node(_null))
+        if len(nodes) == 0 or all(node.value is _null for node in nodes):
+            raise ValueError("Requested level not present in tree.")
+        levels[currlevel] = nodes
+    if show_values is True:
+        levels[level] = [node.value for node in levels[level]]
+    return levels[level]
+
+
+def get_levels(bt, show_values=False, show_nulls=False):
```
Comment (joowani) [inline binarytree/__init__.py]: I think returning a list of lists would be better here. Returning a dict with index as keys is computationally more expensive to create with no additional benefit.

File: binarytree/__init__.py [hunk 5]
```diff
@@ -6,28 +6,41 @@
 def no_op(value):  # pragma: no cover
     return value
 
+
```
[the remark this answers is not in the document]
Comment (cshaley) [reply binarytree/__init__.py]: flake8 said to put a newline there.

File: binarytree/__init__.py [hunk 6]
```diff
@@ -92,11 +126,13 @@ def _right_of(node):
 
 def _set_left(node, child):
     """Set the child to the left of the node."""
+    child.parent = node
```
[the remark this answers is not in the document]
Comment (cshaley) [reply binarytree/__init__.py]: I think so too - I'll test it out before changing to make sure.

File: binarytree/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joowani) [reply binarytree/__init__.py]: I see. That's fine then.

File: binarytree/__init__.py (hunk 1 above)
Comment (cshaley) [reply binarytree/__init__.py]: So I can just say `bt = _prepare_tree(bt)`, correct?

**Comment by cshaley** [comment]
@joowani I believe I made all of the requested changes.  Please review when you get a chance.

**Comment by joowani** [comment]
@cshaley awesome! Thanks again for your contribution.

**Pull request merged**